### ISSUE TITLE
Known issue: Error when trying to use VCP - Didn't find key "remote" in input

### ISSUE THREAD

**Issue by x0b** [x0b/rcx#129]
#### What version of RCX are you using (About -> App version)?
version: ` 1.12.0-r210418`
versionCode: `186`
#### What is your Android version, phone model and manufacturer?
Android 10

#### Which steps are required to reproduce this issue?

1. Enable VCP.
2. Try to browse my files with the Android 10 "Files" app.
3. Receive error `Didn't find key "remote" in input.`

#### What are the contents of ```Android/data/io.github.x0b.rcx/files/logs/log.txt```?
N/A
<details><summary><code>log.txt</code> (click to expand) </summary><br><pre>

</pre></details>

**Comment by alscaldas** [comment]
> > I'm sorry to reopen a closed ticket but I'm having this very same issue.
> 
> No problem, I was pretty much expecting this.
> 
> Since you are already here, can you help me to understand how this happens? The initial fix was based on the following assumptions:
> 
>     1. A user starts RCX and enables the Content Provider Preview (VCP).
> 
>     2. The Content Provider Preview is started and loads the currently configured remotes.
> 
>     3. The user configures a new remote.
> 
>     4. The user tries to use the new remote with the Content Provider preview.
>        warning Since the remote did not exist when rclone was initialized previously, it does not know the new remotes => error message.
> 
> 
> Is that accurate? Or were you doing something else entirely?

Thank you for your prompt reply.

In fact I supposed that once the VCP option was previously enabled on RCX I had to do nothing else in order to get it working.

So what I do is:

I open an app (in this case, Keepass2Android) and I know that RCX works properly because I see an icon on Android tray stating that "Rclone is running in background / No active jobs."

I see the option RCX - Rclone for Android and the total available remotes among the other options I have.
 
When I choose the desired remote (in my case, a crypt remote as previously stated) it shows the error message.

**Comment by x0b** [comment]
1. Do you happen to have a log file? The `Android/data/io.github.x0b.rcx/files/logs/rcd.log` would be really useful.
2. Does the error occur repeatedly or did it only happen once?

I currently don't have a way of reproducing the issue on a test device. I'm currently thinking about adding this error to appcenter and maybe figure out what happens that way.

**Comment by alscaldas** [comment]
1. Sure, the required log file follows attached. [rcd.log](https://github.com/x0b/rcx/files/6469362/rcd.log)
2. It occurs all times I try to use it. I never successfully used this feature so far.

**Comment by ketacean** [comment]
I have exactly the same problem on Android 10 with rcx 1.12.0 downloaded from github's release. It happens every time on every remote.
 Here is my log: 
2021/05/12 22:21:28 DEBUG : Using config file from "/data/user/0/io.github.x0b.rcx/files/rclone.conf"
2021/05/12 22:21:28 DEBUG : rclone: Version "v1.55.1-rcx" starting with parameters ["/data/app/io.github.x0b.rcx-A292E--yIYF2crT3KeKQxQ==/lib/arm64/librclone.so" "--config" "/data/user/0/io.github.x0b.rcx/files/rclone.conf" "--rc-addr" "localhost:42227" "--rc-user" "admin" "--rc-pass" "PrH4eN0RzAVUTe-MB6v21Q==" "--rc-serve" "--log-file" "/storage/emulated/0/Android/data/io.github.x0b.rcx/files/logs/rcd.log" "--dump" "headers" "-vvv" "rcd"]
2021/05/12 22:21:28 INFO  : Using --user admin --pass XXXX as authenticated user
2021/05/12 22:21:28 NOTICE: Serving remote control on http://localhost:42227/
2021/05/12 22:21:39 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:21:39 ERROR : rc: "operations/list": error: Didn't find key "remote" in input
2021/05/12 22:21:39 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:21:39 ERROR : rc: "operations/list": error: Didn't find key "remote" in input
2021/05/12 22:27:49 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:27:49 ERROR : rc: "operations/list": error: Didn't find key "remote" in input
2021/05/12 22:27:54 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:27:54 ERROR : rc: "operations/list": error: Didn't find key "remote" in input
2021/05/12 22:27:55 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:27:55 ERROR : rc: "operations/list": error: Didn't find key "remote" in input
2021/05/12 22:27:56 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:27:56 ERROR : rc: "operations/list": error: Didn't find key "remote" in input
2021/05/12 22:27:58 DEBUG : rc: "operations/list": with parameters map[]
2021/05/12 22:27:58 ERROR : rc: "operations/list": error: Didn't find key "remote" in input

**Comment by ketacean** [comment]
[rcd.log](https://github.com/x0b/rcx/files/6469378/rcd.log)

**Comment by x0b** [comment]
**TLDR**: I figured it out. It's a build/packaging error. VCP is completely non-functional but will be fixed trivially in v1.12.1.

So initially, I assumed that this was a race condition, that something was changing the config file, and stale data would cause the issue. Why? Because that was the actual issue of the old bug, and the old fix corrected this. However, as I dug deeper and deeper into this, it became clear that rclone is actually quite robust against the potential race condition. Also, provoking the race condition resulted in a different error, so this couldn't be the problem.

So I started to inject various faults into the process, until I found one that resulted in the same error message - `Didn't find key "remote" in input.` The only problem: this issue only occured when I completely removed the `remote` key from the class definition. In practice, this should be impossible though, since the definitions are baked into RCX and cannot be changed or configured. And it is. So the fault must be introduced at an earlier point. Inspecting the release APK shows that the code is just...missing:

```diff
diff --git a/ListRcOpParam-expected.class-txt b/ListRcOpParam-actual.class-txt
index 6d387be..4c4d34a 100644
--- a/ListRcOpParam-expected.class-txt
+++ b/ListRcOpParam-actual.class-txt
@@ -1,9 +1,5 @@
 private static class ListRcOpParam implements RcloneRcd.RcOpParam {
-    String fs;
-    String remote;

-    public ListRcOpParam(String fs, String remote) {
-        this.fs = fs;
-        this.remote = remote;
+    public ListRcOpParam() {
     }
 }

#

```

<details><summary>bytecode difference</summary>


```diff
diff --git a/ListRcOpParam-expected.class-txt b/ListRcOpParam-actual.class-txt
index 825ba60..1e19aff 100644
--- a/ListRcOpParam-expected.class-txt
+++ b/ListRcOpParam-actual.class-txt
@@ -5,21 +5,13 @@
 .implements Lca/pkay/rcloneexplorer/RcloneRcd$RcOpParam;


-# instance fields
-.field fs:Ljava/lang/String;
-
-.field remote:Ljava/lang/String;
-
-
 # direct methods
 .method public constructor <init>(Ljava/lang/String;Ljava/lang/String;)V
     .registers 3

     invoke-direct {p0}, Ljava/lang/Object;-><init>()V

-    iput-object p1, p0, Lca/pkay/rcloneexplorer/RcloneRcd$ListRcOpParam;->fs:Ljava/lang/String;
-
-    iput-object p2, p0, Lca/pkay/rcloneexplorer/RcloneRcd$ListRcOpParam;->remote:Ljava/lang/String;
-
     return-void
 .end method
+
+
```
</details>

How does that happen? Well, RCX uses D8/R8 to optimize Java code. Eliminate unused operations, desugar lambdas, etc. And since we use a reflection-based JSON library, it can be hard to infer which code is used and which is not. This is precisely why this issue already popped up during the development of VCP, and I added a `-keep` rule to disable D8/R8 for the model classes. For some yet unknown reason, this rule was not applied while building the release APK for  `v1.12.0`. 

I'm still not sure why - I discovered this issue while preparing a release APK with extended logging for you, and R8 did _not_ remove the model members 🤷‍♀️. In any case, this also highlights that RCX really needs some automated tests 🙈.

**Comment by x0b** [comment]
This should be fixed in v1.12.1. I'll leave this open for now in case the error pops up again.

**Comment by alscaldas** [comment]
Just to let you know, this new release seems to have fixed this trouble for me. I've been running Keepass2Android and now I'm able to open a kdbx file from one of my Rclone remotes.

Much appreciated.
